File: src/io/io.cc
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

// Copyright (c) 2015 by Contributors

#include <mxnet/io.h>
#include <dmlc/registry.h>
#include "./image_augmenter.h"
#include "./image_iter_common.h"

// Registers
namespace dmlc {
DMLC_REGISTRY_ENABLE(::mxnet::DataIteratorReg);
DMLC_REGISTRY_ENABLE(::mxnet::DatasetReg);
DMLC_REGISTRY_ENABLE(::mxnet::BatchifyFunctionReg);
}  // namespace dmlc

namespace mxnet {
namespace io {
// Register parameters in header files
DMLC_REGISTER_PARAMETER(BatchParam);
DMLC_REGISTER_PARAMETER(BatchSamplerParam);
DMLC_REGISTER_PARAMETER(PrefetcherParam);
DMLC_REGISTER_PARAMETER(ImageNormalizeParam);
DMLC_REGISTER_PARAMETER(ImageRecParserParam);
DMLC_REGISTER_PARAMETER(ImageRecordParam);
DMLC_REGISTER_PARAMETER(ImageDetNormalizeParam);
}  // namespace io
}  // namespace mxnet

File: scala-package/core/src/main/scala/org/apache/mxnet/javaapi/IO.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.mxnet.javaapi
import scala.language.implicitConversions

class DataDesc private[mxnet] (val dataDesc: org.apache.mxnet.DataDesc) {

  def this(name: String, shape: Shape, dType: DType.DType, layout: String) =
    this(new org.apache.mxnet.DataDesc(name, shape, dType, layout))

  override def toString(): String = dataDesc.toString()
}

object DataDesc{
  implicit def fromDataDesc(dDesc: org.apache.mxnet.DataDesc): DataDesc = new DataDesc(dDesc)

  implicit def toDataDesc(dataDesc: DataDesc): org.apache.mxnet.DataDesc = dataDesc.dataDesc

  /**
    * Get the dimension that corresponds to the batch size.
    * @param layout layout string. For example, "NCHW".
    * @return An axis indicating the batch_size dimension. When data-parallelism is used,
    *         the data will be automatically split and concatenate along the batch_size dimension.
    *         Axis can be -1, which means the whole array will be copied
    *         for each data-parallelism device.
    */
  def getBatchAxis(layout: String): Int = org.apache.mxnet.DataDesc.getBatchAxis(Some(layout))
}
